[Filters] Manage applying the CSSFilter using FilterTargetSwitcher

shallawa opened the pull request:
#### 644b784fd5f7d09ad596d31a1c8af4f01fb065cd
<pre>
[Filters] Manage applying the CSSFilter using FilterTargetSwitcher
<a href="https://bugs.webkit.org/show_bug.cgi?id=248175">https://bugs.webkit.org/show_bug.cgi?id=248175</a>
rdar://102585518

Reviewed by Simon Fraser.

FilterTargetSwitcher can prepare the drawing context and can apply the filter when
the source image drawing is complete. So it will make since if we unify applying
the CSSFilter, SVGFilter and canvas filter by using this class.

This patch will use FilterTargetSwitcher to apply the CSSFilter. In future patches
SVGFilter and canvas filter will use it as well.

For applying software filters we will use a variant of FilterTargetSwitcher named
FilterImageTargetSwitcher. This class will manage creating the source image and
maintaining the filter results. A new super class of FilterTargetSwitcher will be
introduced in future patches for the CoreGraphics filters.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/filters/FilterImageTargetSwitcher.cpp: Copied from Source/WebCore/platform/graphics/filters/FilterTargetSwitcher.h.
(WebCore::FilterImageTargetSwitcher::FilterImageTargetSwitcher):
(WebCore::FilterImageTargetSwitcher::drawingContext const):
(WebCore::FilterImageTargetSwitcher::didDrawSourceImage):
* Source/WebCore/platform/graphics/filters/FilterImageTargetSwitcher.h: Copied from Source/WebCore/platform/graphics/filters/FilterTargetSwitcher.h.
* Source/WebCore/platform/graphics/filters/FilterTargetSwitcher.cpp:
(WebCore::FilterTargetSwitcher::create):
(WebCore::FilterTargetSwitcher::FilterTargetSwitcher):
(WebCore::FilterTargetSwitcher::willDrawSourceImage):
(WebCore::FilterTargetSwitcher::didDrawSourceImage):
(WebCore::FilterTargetSwitcher::~FilterTargetSwitcher): Deleted.
(WebCore::FilterTargetSwitcher::drawingContext const): Deleted.
(WebCore::FilterTargetSwitcher::outsets const): Deleted.
* Source/WebCore/platform/graphics/filters/FilterTargetSwitcher.h:
(WebCore::FilterTargetSwitcher::drawingContext const):
(WebCore::FilterTargetSwitcher::needsRedrawSourceImage):
* Source/WebCore/rendering/RenderLayerFilters.cpp:
(WebCore::RenderLayerFilters::beginFilterEffect):
(WebCore::RenderLayerFilters::applyFilterEffect):
(WebCore::RenderLayerFilters::inputContext): Deleted.
(WebCore::RenderLayerFilters::allocateBackingStoreIfNeeded): Deleted.
* Source/WebCore/rendering/RenderLayerFilters.h:

Canonical link: <a href="https://commits.webkit.org/257029@main">https://commits.webkit.org/257029@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f30e29a790b6668635ff9980e08071978daf5a9f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/97650 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/6896 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/30816 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/107147 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/167410 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/101598 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/7260 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/35658 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/90046 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/103802 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/103295 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/5436 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/84281 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/32448 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/87316 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/89106 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/75368 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/894 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/20553 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/884 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/22036 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4838 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/5701 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/44498 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/2134 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/41428 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/platform/graphics/filters/FilterImageTargetSwitcher.h:
```diff
@@ -0,0 +1,49 @@
+
+class ImageBuffer;
+
+class FilterImageTargetSwitcher : public FilterTargetSwitcher {
```
`final`

smfr commented on Source/WebCore/platform/graphics/filters/FilterImageTargetSwitcher.h:
```diff
@@ -0,0 +1,49 @@
+
+class FilterImageTargetSwitcher : public FilterTargetSwitcher {
+    WTF_MAKE_FAST_ALLOCATED;
+
```
No blank line here.

smfr commented on Source/WebCore/rendering/RenderLayerFilters.cpp:
```diff
@@ -185,44 +172,34 @@ GraphicsContext* RenderLayerFilters::beginFilterEffect(RenderElement& renderer,
-    auto* sourceGraphicsContext = inputContext();
-    if (!sourceGraphicsContext)
+    if (!m_targetSwitcher || hasUpdatedBackingStore)
+        m_targetSwitcher = FilterTargetSwitcher::create(context, filter, filterRegion, DestinationColorSpace::SRGB());
```
Do we need to heap allocate this, or can it live on the stack in the caller, and get passed in? That way the lifetime is much clearer.

shallawa replied:
The plan is to add a superclass of FilterTargetSwitcher named FilterStyleTargetSwitcher for applying the filters using GraphicsContext filters. So yes we need to keep this heap allocated.

shallawa committed:
[Filters] Manage applying the CSSFilter using FilterTargetSwitcher
https://bugs.webkit.org/show_bug.cgi?id=248175
rdar://102585518

Reviewed by Simon Fraser.

FilterTargetSwitcher can prepare the drawing context and can apply the filter when
the source image drawing is complete. So it will make since if we unify applying
the CSSFilter, SVGFilter and canvas filter by using this class.

This patch will use FilterTargetSwitcher to apply the CSSFilter. In future patches
SVGFilter and canvas filter will use it as well.

For applying software filters we will use a variant of FilterTargetSwitcher named
FilterImageTargetSwitcher. This class will manage creating the source image and
maintaining the filter results. A new super class of FilterTargetSwitcher will be
introduced in future patches for the CoreGraphics filters.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/filters/FilterImageTargetSwitcher.cpp: Copied from Source/WebCore/platform/graphics/filters/FilterTargetSwitcher.h.
(WebCore::FilterImageTargetSwitcher::FilterImageTargetSwitcher):
(WebCore::FilterImageTargetSwitcher::drawingContext const):
(WebCore::FilterImageTargetSwitcher::didDrawSourceImage):
* Source/WebCore/platform/graphics/filters/FilterImageTargetSwitcher.h: Copied from Source/WebCore/platform/graphics/filters/FilterTargetSwitcher.h.
* Source/WebCore/platform/graphics/filters/FilterTargetSwitcher.cpp:
(WebCore::FilterTargetSwitcher::create):
(WebCore::FilterTargetSwitcher::FilterTargetSwitcher):
(WebCore::FilterTargetSwitcher::willDrawSourceImage):
(WebCore::FilterTargetSwitcher::didDrawSourceImage):
(WebCore::FilterTargetSwitcher::~FilterTargetSwitcher): Deleted.
(WebCore::FilterTargetSwitcher::drawingContext const): Deleted.
(WebCore::FilterTargetSwitcher::outsets const): Deleted.
* Source/WebCore/platform/graphics/filters/FilterTargetSwitcher.h:
(WebCore::FilterTargetSwitcher::drawingContext const):
(WebCore::FilterTargetSwitcher::needsRedrawSourceImage):
* Source/WebCore/rendering/RenderLayerFilters.cpp:
(WebCore::RenderLayerFilters::beginFilterEffect):
(WebCore::RenderLayerFilters::applyFilterEffect):
(WebCore::RenderLayerFilters::inputContext): Deleted.
(WebCore::RenderLayerFilters::allocateBackingStoreIfNeeded): Deleted.
* Source/WebCore/rendering/RenderLayerFilters.h:

Canonical link: https://commits.webkit.org/257029@main
